Support for parameters with arguments

rzikm opened the pull request:
Was playing around with gitu and wanted to refresh some of the Rust I know. This is a very rough sketch which allows me to add arguments to flags (such as `git log -n256` or in the future for `git log -S`).

It's far from optimal, or production quality, just wanted to float it out there and see if you agree with the direction I took. We can iterate from there.

altsem commented on src/ops/log.rs:
```diff
@@ -44,8 +47,23 @@ fn log_other(state: &mut State, _term: &mut Term, _args: &[OsString], result: &s
+    let limit = state
+        .pending_menu
+        .as_ref()
+        .map(|m| m.args.get("-n"))
+        .flatten()
+        .map(Arg::get_i32)
+        .flatten()
```
I *think* this results in the same behaviour!
```suggestion
        .and_then(|m| m.args.get("-n"))
        .and_then(Arg::get_i32)
```

altsem reviewed (changes requested):
:+1:
I think it looks pretty good! Happy to see you pick this up, arguments with values is kind of a missing feature.

See if the comments make any sense.

rzikm commented on src/menu/arg.rs:
```diff
@@ -1,24 +1,100 @@
+#[derive(Debug, Clone)]
+pub(crate) enum ArgValue {
+    Bool(bool),
+    String(String),
```
> (could the bool even be Option<()>? I don't know if that's crazy, or genuis :D)

I was thinking about it, but I don't see what this improves

For way into the future, I think there should be something like "ArgValueDescriptor", which would contain information such as type, list of valid values (for choice parameters, such as `git diff --diff-algorithm`), but I am not familiar with how these things are usually done in Rust and am not willing to fight the compiler over it.

altsem replied:
I got it wrong (as with the comment below about the `Option`.
`ArgValueDescriptor` might make sense.

rzikm commented on src/menu/arg.rs:
```diff
@@ -1,24 +1,100 @@
     pub arg: &'static str,
     pub display: &'static str,
-    pub(crate) state: bool,
+    pub(crate) value: ArgValue,
```
We need two pieces of information per Arg:
- type of the parameter
- actual value of the parameter.

If value is `None`, we lose the knowledge of the type of the Arg, unless I am missing something.

altsem replied:
Ah no, you're right!

altsem commented:
@rzikm Code looks good I think. Just some clippy complaints, "feat: " missing, tests.

altsem commented on src/ops/log.rs:
```diff
@@ -4,7 +4,10 @@ use derive_more::Display;
-pub(crate) const ARGS: &[Arg] = &[];
+pub(crate) const ARGS: &[Arg] = &[
+    Arg::new_int_opt("-n", "Limit number of commits", Some(256)),
+    // Arg::new_str("-S", "Search occurences"), // TOOD: Implement search
```
Just a reminder. `-S` is added in `src/default_config.toml`, not sure if it'd be visible in the UI. But lets either include the keybinding & impl, or none.

rzikm commented:
Did some work on this, I am not happy with how the `Arg` structure turned out, the fact that we need the Arg::new* functions be complicates it a bit. Maybe if we make it into a `get_items()` instead, I could simplify `Arg` a lot.

Anyway, it turns out this change greatly speeds up log screen because we no longer try to read the entire history :D very noticeable on Windows on large repos.

I removed the -S parameter for now, and added -F instead so that it can be checked that the shape works for other arg types.

rzikm commented:
Should be good for another review

altsem commented:
@rzikm I've been busy with work recently, I'll try find some time soon to look at it.

altsem reviewed (commented):
I dug around a bit. I managed to remove the need to implement Clone! :) It's over at the branch [altsem/value-arguments](https://github.com/altsem/gitu/commits/altsem/value-arguments/).

I had some idea that the tests could:
- Create some commits (there's a `commit` helper function).
- Limit log to a number less than these commits (We've tested that limiting somewhat works).

Same with grepping:
- Create two commits, e.g. "Commit 1", "Another commit"
- Grep for "Another"
- Only one should be in the log!

I think it would be nice to flatten a few structs maybe. Could dig around a lot. But at some point I'm all for merging this into master and take it from there, it doesn't have to be perfect initially :)

rzikm commented:
Thanks a lot, now it looks reasonable to me,

I added some tests as asked.

altsem reviewed (changes requested):
I'm generally happy with the implementation.
Just found one problem: `--grep` when doing `log other`.

I tried to make `src/tests/log.rs` a bit more clear over at `altsem/value-arguments`. Also added a test for the above.

altsem commented:
I haven't exactly grasped what's going on. pending_menu (and state of the arguments) should exist until the "log other" rev prompt is done and "log other" callback has been executed I think.

Also found that grepping for something that does not match seems to cause a panic:
```sh
cargo run -- -k 'l-Fxxxxxxxxx<enter>l'
```

altsem commented:
@rzikm Figured I'd share. After some digging I think I found an approach that might work.
The idea is to:
- Remove the `args: [&OsString]` parameter from prompt callbacks.
- Hide the menu when a prompt is active (you can still access its args).
- Call `state.close_menu()` explicitly in each action/prompt.

This way we can close the prompt where appropriate (some actions don't), and after the args have been assigned to commands.

My work so far: [https://github.com/altsem/gitu/commit/99a75bbe69cb49342510c2a6d2d7b215679464fa](https://github.com/altsem/gitu/commit/99a75bbe69cb49342510c2a6d2d7b215679464fa)

Heading for vacay and might continue next week :)

rzikm commented:
Thanks, I did not have time to spend on this last couple of days, hopefully I will get to it this week so we can finally draw this to finish line.

rzikm commented:
I am sorry for dragging this, I was busy last couple of weeks, I hope to get to this soon

altsem commented:
That's alright, no need to stress.

altsem commented:
I continued on the branch: [altsem/value-arguments](https://github.com/altsem/gitu/tree/altsem/value-arguments)

Made a test for grepping for something that doesn't exist. It seems that we get a crash when showing a screen with no lines.

Otherwise I think I solved the other issues :)

rzikm commented:
I fixed the failing test by ensuring there is always at least one item in the log screen, I did not want to start rewriting the screen logic to allow for empty item list.

Let me know if there is anything more missing.

altsem commented:
I think we're ready. Good job! :)

I did a rebase onto master, and will do a hack to add some entries to the changelog. Will make a release soon :raised_hands:

altsem commented:
Pushed here with a rebase instead / some empty commits just for changelog entries purposes:
https://github.com/altsem/gitu/pull/195
